File: ios/MegaPDF/RecentLocation.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Where a recent document lives, as the Files app would say it (#165).
///
/// A recent list shows file names, and files from one template or one scanner
/// share them: three rows reading "agreement.pdf" tell nobody which is which.
/// What separates them is the folder, so every row carries its location — not
/// only the rows whose names clash, which would make the list jump about as
/// entries come and go.
///
/// Recorded when the file is opened, not derived when the list is drawn. A
/// `RecentEntry` holds a security-scoped bookmark and nothing else; resolving ten
/// bookmarks and taking security-scoped access on each of them to label a list is
/// the wrong thing to do while a view is being laid out. At the moment a document
/// is opened the app has the URL, and the right to read around it, which is
/// exactly what this needs.
///
/// iOS shows no paths, so neither does this: display names only, at most two of
/// them, in the app's language.
struct RecentLocation: Codable, Equatable, Hashable, Sendable {
    /// Outermost first, at most two: `["iCloud Drive", "Clients"]`, `["Downloads"]`.
    /// Never a path, a container identifier or a `file://` URL.
    let segments: [String]

    /// The line under the file name. Finder's path bar and the Files app both
    /// separate places with "›", and the Windows and Mac halves of #165 use it too.
    var subtitle: String { segments.joined(separator: " › ") }

    var isEmpty: Bool { segments.isEmpty }
}

extension RecentLocation {

    // MARK: - the rules

    /// What the file system said about a file. Everything the naming rules are
    /// allowed to use, gathered in one place so that the rules can be tested
    /// without an iCloud account, a file provider or a device.
    struct Facts: Equatable, Sendable {
        /// `URLResourceKey.isUbiquitousItemKey`: the item is in iCloud.
        var isUbiquitous = false
        /// `ubiquitousItemContainerDisplayNameKey`, when iCloud gave one.
        var iCloudContainerName: String?
        /// The parent folder's `localizedName`, else its last path component.
        var parentName: String?
        /// The file is inside this app's own container, so it is on the device.
        var isInAppContainer = false
        /// "On My iPhone" or "On My iPad", localised.
        var deviceName: String = ""

        init(isUbiquitous: Bool = false, iCloudContainerName: String? = nil,
             parentName: String? = nil, isInAppContainer: Bool = false,
             deviceName: String = "") {
            self.isUbiquitous = isUbiquitous
            self.iCloudContainerName = iCloudContainerName
            self.parentName = parentName
            self.isInAppContainer = isInAppContainer
            self.deviceName = deviceName
        }
    }

    /// Folder names that are the file system's business rather than a person's.
    /// A file provider keeps its files under "File Provider Storage" inside an app
    /// group whose folder is a UUID, and iCloud's own folder is
    /// "com~apple~CloudDocs". None of those is a place anybody would recognise,
    /// and the spec for this is explicit that a row never shows container paths.
    private static let internalFolderNames: Set<String> = [
        "File Provider Storage", "Mobile Documents", "Containers", "AppGroup", "Data",
    ]

    /// The folder name to show, or nil when the name is the file system's and not
    /// the person's. A folder called "Documents" or "Downloads" is a real place and
    /// stays; one called "com~apple~CloudDocs" or a bare UUID does not.
    static func presentableFolderName(_ raw: String?) -> String? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines),
              !trimmed.isEmpty,
              trimmed != "/",
              !trimmed.contains("~"),
              !trimmed.contains("/"),
              UUID(uuidString: trimmed) == nil,
              !internalFolderNames.contains(trimmed)
        else { return nil }
        return trimmed
    }

    /// The location a row shows, from what the file system said.
    ///
    /// The place comes first when there is one to name: iCloud says so itself, and
    /// a file inside this app's own container is on the device. A file reached
    /// through some other provider cannot be named — iOS offers no public way to
    /// turn a file provider's domain identifier into "Dropbox" — so rather than
    /// guess, such a row shows its folder alone, which is the part that tells two
    /// same-named files apart anyway.
    static func from(_ facts: Facts) -> RecentLocation {
        var place: String?
        if facts.isUbiquitous {
            let named = facts.iCloudContainerName?.trimmingCharacters(in: .whitespacesAndNewlines)
            place = (named?.isEmpty == false ? named : nil) ?? String(localized: "iCloud Drive")
        } else if facts.isInAppContainer {
            place = facts.deviceName.isEmpty ? nil : facts.deviceName
        }

        var segments: [String] = []
        if let place { segments.append(place) }
        // A file at the root of its place has a parent the system names after the
        // place itself ("iCloud Drive"); it adds nothing, so it is left off.
        if let folder = presentableFolderName(facts.parentName),
           folder.caseInsensitiveCompare(place ?? "") != .orderedSame {
            segments.append(folder)
        }
        return RecentLocation(segments: segments)
    }

    // MARK: - reading the facts

    /// "On My iPhone" or "On My iPad", the way the Files app labels the device.
    static func deviceName() -> String {
        #if canImport(UIKit)
        if UIDevice.current.userInterfaceIdiom == .pad {
            return String(localized: "On My iPad",
                          comment: "#165: where a recent file lives; the Files app's name for this iPad")
        }
        #endif
        return String(localized: "On My iPhone",
                      comment: "#165: where a recent file lives; the Files app's name for this iPhone")
    }

    /// Reads the facts for a file that is open.
    ///
    /// The caller must already hold security-scoped access — `ViewerModel` holds it
    /// for the document's whole life — because the parent folder's display name is
    /// outside the sandbox extension the picker grants otherwise. Every read is
    /// best-effort: a folder name that cannot be read falls back to the last path
    /// component, and a row with no location at all is better than a wrong one.
    ///
    /// `deviceName` is passed in rather than read here: this runs off the main actor,
    /// away from the view being drawn, and `UIDevice` belongs on the main one.
    static func facts(for url: URL, deviceName: String) -> Facts {
        var facts = Facts()
        facts.deviceName = deviceName
        facts.isInAppContainer = url.standardizedFileURL.path
            .hasPrefix(URL(fileURLWithPath: NSHomeDirectory()).standardizedFileURL.path)

        let values = try? url.resourceValues(
            forKeys: [.isUbiquitousItemKey, .ubiquitousItemContainerDisplayNameKey])
        facts.isUbiquitous = values?.isUbiquitousItem ?? false
        facts.iCloudContainerName = values?.ubiquitousItemContainerDisplayName

        let parent = url.deletingLastPathComponent()
        facts.parentName =
            (try? parent.resourceValues(forKeys: [.localizedNameKey]))?.localizedName
            ?? parent.lastPathComponent.removingPercentEncoding
            ?? parent.lastPathComponent
        return facts
    }

    /// The location to store for a file being opened, or nil when nothing about it
    /// can be named. Never throws: a recent entry without a location is a row
    /// without a subtitle, not a failure to open a document.
    static func of(_ url: URL, deviceName: String) -> RecentLocation? {
        let location = from(facts(for: url, deviceName: deviceName))
        return location.isEmpty ? nil : location
    }
}

File: android/app/src/main/java/com/megapdf/android/RecentLocation.kt
package com.megapdf.android

/**
 * Where a recent document lives, in the words the system file picker uses (#165).
 *
 * The Windows half of #165 has `RecentLocation.cs` for the same job; this is the
 * Android one, and it is deliberately the same shape — pure string work with no
 * Android types in it, so the awkward cases are tested on the JVM rather than on
 * an emulator.
 *
 * **What Android will and will not tell us.** `ACTION_OPEN_DOCUMENT` hands back a
 * *document* URI, not a tree, so there is no parent to walk: `DocumentsContract`
 * has no "give me this document's folder" call. What there is:
 *
 * - the **root** the document came from. Not its *title*, which only the picker
 *   may read (see `DocumentLocations`), but a name for it that needs no
 *   permission: the volume's description, or the owning app's label;
 * - the **document id**, whose shape is the provider's business. Exactly one
 *   provider documents it, and it is the common one: external storage uses
 *   `<rootId>:<relative/path>`, so its folders can be read straight off.
 *
 * So a row shows the root always, and the folders underneath it when the provider
 * spells them out. "Downloads", "Drive", "Internal storage › Documents › Clients".
 * That is what the picker itself shows, which is the point.
 *
 * Nothing here ever emits a `content://` URI, a `/storage/emulated/0/…` path or a
 * document id: those are the three things #165 says never to put in front of a
 * person, and [segments] is the only way out of this file.
 */
object RecentLocation {

    /** The separator #165 uses on every platform. */
    const val SEPARATOR = " › "

    /** The one provider whose document ids are documented to carry a path. */
    const val EXTERNAL_STORAGE = "com.android.externalstorage.documents"

    /** The provider whose root MegaPDF names itself — see `DocumentLocations`. */
    const val DOWNLOADS = "com.android.providers.downloads.documents"

    /**
     * The folders to show for a document, outermost first.
     *
     * [rootTitle] is what the provider calls the root — already localised by it,
     * because a provider publishes its title in the device's language. [documentId]
     * is the provider's own id for the document, and is only read when the provider
     * is one whose ids are documented to be paths.
     */
    fun segments(authority: String?, documentId: String?, rootTitle: String?): List<String> {
        val root = rootTitle?.trim().orEmpty()
        val folders = if (authority == EXTERNAL_STORAGE) foldersFromPathId(documentId) else emptyList()
        return (listOf(root) + folders).filter { it.isNotEmpty() }
    }

    /**
     * The folders inside `<rootId>:<relative/path/name.pdf>`, without the file name.
     *
     * A document at the root of the volume has no folders, and an id in any other
     * shape gets none — better a row that says only "Internal storage" than one
     * that leaks half a path.
     */
    private fun foldersFromPathId(documentId: String?): List<String> {
        val id = documentId ?: return emptyList()
        val colon = id.indexOf(':')
        if (colon < 0) return emptyList()
        val relative = id.substring(colon + 1)
        // An absolute path here would mean the provider changed its id format under
        // us; drop it rather than print it.
        if (relative.startsWith("/")) return emptyList()
        return relative.split('/')
            .dropLast(1)          // the file itself
            .filter { it.isNotBlank() }
    }

    /**
     * [segments] with any root *we* named swapped for its name in the language the
     * app is in now (#165).
     *
     * A location is recorded once, when the document is opened, and folders on
     * disk keep whatever they are called — "Clients" is "Clients" in French. But
     * "Downloads" is a word of ours, and a stored one goes stale the moment
     * someone changes the app's language, so it is re-read rather than replayed.
     */
    fun localised(segments: List<String>, authority: String?, downloadsName: String): List<String> =
        if (authority == DOWNLOADS && segments.isNotEmpty())
            listOf(downloadsName) + segments.drop(1)
        else segments

    /**
     * [segments] with its root replaced by [root], which is the same name asked for
     * again in the language the app is in now (#165).
     *
     * Nothing changes when the root cannot be resolved: what was recorded at open
     * time is still the truth about where the file is, just possibly in the wrong
     * language, and that beats a row with no location at all.
     */
    fun withRoot(segments: List<String>, root: String?): List<String> {
        val name = root?.trim().orEmpty()
        if (name.isEmpty() || segments.isEmpty()) return segments
        return listOf(name) + segments.drop(1)
    }

    /**
     * [segments] as one line, shortened to at most [maxSegments] by dropping from
     * the middle.
     *
     * The root and the innermost folders survive, because those are the two ends
     * that tell two same-named files apart: the root says which provider, and the
     * last folder is usually the answer to "which of these two?". #165 says the
     * same thing for Windows — "never away from the folder that tells two entries
     * apart".
     */
    fun format(segments: List<String>, maxSegments: Int = 3, ellipsis: String = "…"): String {
        if (segments.isEmpty()) return ""
        if (maxSegments < 1) return ""
        if (segments.size <= maxSegments) return segments.joinToString(SEPARATOR)
        if (maxSegments == 1) return segments.last()
        if (maxSegments == 2) return segments.first() + SEPARATOR + segments.last()
        val tail = segments.takeLast(maxSegments - 2)
        return (listOf(segments.first(), ellipsis) + tail).joinToString(SEPARATOR)
    }
}
